api/turing/api/ensemblers_api.go

```diff
@@ -109,18 +111,160 @@ func (c EnsemblersController) UpdateEnsembler(
 			"Ensembler type cannot be changed after creation")
 	}
 
+	pyFuncEnsembler, isPyfunc := ensembler.(*models.PyFuncEnsembler)
+
+	var oldPyFuncEnsembler models.PyFuncEnsembler
+	if isPyfunc {
+		oldPyFuncEnsembler = models.PyFuncEnsembler{
+			RunID:       pyFuncEnsembler.RunID,
+			ArtifactURI: pyFuncEnsembler.ArtifactURI,
+		}
+	}
+
 	if err = ensembler.Patch(request.EnsemblerLike); err != nil {
 		return BadRequest("invalid ensembler configuration", err.Error())
 	}
-
 	ensembler, err = c.EnsemblersService.Save(ensembler)
 	if err != nil {
+		// Delete If Only RunID is changed, if only name is changed, doesn't need to delete the mlflow run
+		if isPyfunc && oldPyFuncEnsembler.RunID != pyFuncEnsembler.RunID {
```

Comment:
If the only thing that updated were the names, mlflow wont create another run, so we don't need to delete old and new run, we just need to change the database

> Why are we deleting the old runs if there is an error?

When he update process failed, we delete the new mlflow run, since if we update the ensembler from the sdk, the sdk already create a new mlflow run, so when the update is failed, the new mlflow run will get deleted.

Reply:
> If the only thing that updated were the names

So if I understood correctly, does this refer to the mlflow experiment name? In other words, the ensembler name?

Reply:
yepp, the name refer to ensembler name, but if the ensembler name changed, the mlflow experiment name wont change

Reply:
> We are not checking for the name change (mlflow experiment name) anywhere in this code, right? 

I think what Alex means is that even though we don't check the name change, mlflow will do it implicitly when the SDK calls the necessary mlflow API endpoints to update the experiment. CMIIW, if mlflow detects only a name change, the run id wouldn't change, so we'd need to check to prevent the existing run id from being deleted.

Reply:
> I think what Alex means is that even though we don't check the name change, mlflow will do it implicitly when the SDK calls the necessary mlflow API endpoints to update the experiment. CMIIW, if mlflow detects only a name change, the run id wouldn't change, so we'd need to check to prevent the existing run id from being deleted.

Yes, That's Correct, if the only thing that changed is the name, mlflow won't create a new run, meaning that the runID wont change, so we only need to check if the RunID is changing

Reply:
I see.. Ok. @zenovore, could you add this explanation to the inline comments too? It makes it clearer. Thanks.